src/Transform/ProcessAffineParallelPrivate.cpp

```diff
@@ -0,0 +1,196 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Transform/ProcessAffineParallelPrivate.hpp"
+#include "src/Pass/Passes.hpp"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/Support/Debug.h"
+
+#include "src/Support/TypeUtilities.hpp"
+
+#define DEBUG_TYPE "affine-parallel-private"
+
+using namespace mlir;
+
+namespace {
+func::FuncOp functionBeingDebugged;
+
+struct ProcessAffineParallelWithoutScopePattern
+    : public OpRewritePattern<affine::AffineParallelOp> {
+  using OpRewritePattern<affine::AffineParallelOp>::OpRewritePattern;
+
+  static bool matchParallelForWithAllocScope(
+      affine::AffineParallelOp parForOp) {
+    fprintf(
+        stderr, "\nhi alex, look for parallel for with/without alloca scope\n");
+    parForOp.dump();
+    // auto blockList = parForOp.getRegion().getBlocks();
+    // if (!blockList || blockList->size() == 0) {
+    if (parForOp.getRegion().empty()) {
+      fprintf(stderr, "hi alex, ignore empty parallel region\n");
+      return true;
+    }
+    Block *loopBody = parForOp.getBody();
+    Operation &firstOp = loopBody->front();
+    if (!isa<memref::AllocaScopeOp>(&firstOp)) {
+      fprintf(
+          stderr, "hi alex, found a parallel region without an alloca scope\n");
+      return false;
+    }
+    fprintf(stderr, "hi alex, found a parallel region WITH an alloca scope\n");
+    return true;
+  }
+
+  LogicalResult matchAndRewrite(affine::AffineParallelOp parForOp,
+      PatternRewriter &rewriter) const final {
+    fprintf(stderr, "hi alex, add alloca scope to a parallel region\n");
+    Location loc = parForOp.getLoc();
+    assert(!matchParallelForWithAllocScope(parForOp) &&
+           "expected par for without alloca here");
+#if 0
+    // seems generally bad to clone the op as it has a hard time removing stuff later.
+    auto newOp = rewriter.clone(*parForOp.getOperation());
+    auto newParForOp = cast<affine::AffineParallelOp>(newOp);
+#else
+    SmallVector<Type, 4> resultTypes;
+    for (auto t : parForOp.getResults()) {
+      resultTypes.emplace_back(t.getType());
+    }
+    auto newParForOp = rewriter.create<affine::AffineParallelOp>(loc,
+        resultTypes, parForOp.getReductionsAttr(), parForOp.getLowerBoundsMap(),
+        parForOp.getLowerBoundsGroupsAttr(), parForOp.getUpperBoundsMap(),
+        parForOp.getUpperBoundsGroupsAttr(), parForOp.getSteps(),
+        parForOp.getMapOperands());
+    newParForOp.getRegion().takeBody(parForOp.getRegion());
+#endif
+#if 1
+    // Code inspired from SCFToOpenMP.cpp, in ParallelOpLowering struct, line
+    // 399.
+    {
+      OpBuilder::InsertionGuard allocaGuard(rewriter);
+      // Create a block containing the ops in the loop body.
```

Comment:
I am thinking it's because this ops block and `scopeBlock` are in different regions. I am not sure if it helps if we, instead of `takeBody` at Line 90,  inline the region of the original parForOp into newParForOp, using:
```
rewriter.inlineRegionBefore(parForOp.getRegion(), newParForOp.getRegion(),
                                    newParForOp.getRegion().begin());
```
like SCFToOpenMP: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Conversion/SCFToOpenMP/SCFToOpenMP.cpp#L422

Reply:
Thanks, used your tip, and also had to make a bunch of other dialect legal. It then worked.